switch to moby/sys/{mount,mountinfo} part II

kolyshkin opened the pull request:
This PR finishes the replacement of the `pkg/mount` package with github.com/moby/sys/mount and github.com/moby/sys/mountinfo.

Step by step:
 - [x] Vendor in github.com/moby/sys
 - [x] make pkg/mount a shallow package
(the above is https://github.com/moby/moby/pull/40691; the below is this PR)
 - [x] really switch this repo to use moby/sys directly
 - [x] add deprecation notice to pkg/mount

The long-term plan is to remove `pkg/mount` entirely. This, of course, depends on its external users (I'm afraid we have a lot; I was going to get it from e.g. https://pkg.go.dev/github.com/docker/docker@v1.13.0/pkg/mount?tab=importedby but because it does not recognize moby's v17...v19 tags it's not possible, so it shows users of v.1.13.0. Filed https://github.com/golang/go/issues/37851)

kolyshkin committed:
vndr: vendor in github.com/moby/sys

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
pkg/mount: make it shallow

Make pkg/mount a shallow package, relying on
github.com/moby/sys/mount and github.com/moby/sys/mountinfo

The plan is to
 - switch the rest of this repo to use moby/sys directly
 - add deprecation notice to pkg/mount
 - (eventually) remove pkg/mount

The nolint:golint annotation is added to suppress warnings like
"exported XXX should have comment or be unexported".

The ForceMount function is deprecated (and is identical to Mount),
so let's not use it (but still provide an alias for those who do).

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
Really switch to moby/sys/mount*

Switch to moby/sys/mount and mountinfo. Keep the pkg/mount for potential
outside users.

This commit was generated by the following bash script:

```
set -e -u -o pipefail

for file in $(git grep -l 'docker/docker/pkg/mount"' | grep -v ^pkg/mount); do
	sed -i -e 's#/docker/docker/pkg/mount"#/moby/sys/mount"#' \
		-e 's#mount\.\(GetMounts\|Mounted\|Info\|[A-Za-z]*Filter\)#mountinfo.\1#g' \
		$file
	goimports -w $file
done
```

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
pkg/mount: deprecate

Add a deprecation/removal notice, pointing out to appropriate
replacement packages.

I was not sure if a package-level deprecation is enough, so
I also added notices around each block.

Note that `nolint:golint` annotations are left as is, otherwise
golint complains like this:

> pkg/mount/deprecated.go:45:1: comment on exported var `MergeTmpfsOptions` should be of the form `MergeTmpfsOptions ...` (golint)
> // Deprecated: use github.com/moby/sys/mount instead.
> ^

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
CI failures seem unrelated
> [2020-03-14T01:38:39.122Z] === Failed
> [2020-03-14T01:38:39.122Z] === FAIL: s390x.integration.network TestInspectNetwork (7.44s)
> [2020-03-14T01:38:39.122Z] 
> [2020-03-14T01:38:39.122Z] === FAIL: s390x.integration.network TestServiceWithDataPathPortInit (27.00s)
> [2020-03-14T01:38:39.122Z]     service_test.go:378: timeout hit after 15s: running task count at 0 waiting for 1 (total tasks: 4)
> [2020-03-14T01:38:39.122Z] 
> [2020-03-14T01:38:39.122Z] === FAIL: s390x.integration.network TestServiceWithDefaultAddressPoolInit (17.38s)
> [2020-03-14T01:38:39.122Z]     service_test.go:420: timeout hit after 15s: running task count at 0 waiting for 1 (total tasks: 4)

thaJeztah commented on pkg/mount/flags.go:
```diff
@@ -1,137 +1,34 @@
-	)
+//nolint:golint
+const (
+	RDONLY      = sysmount.RDONLY
```
can you add `// Deprecated: ` comments after each?

Similar to what I did in https://github.com/moby/moby/commit/f572df7c22c1f48b9a2ddd51519797a5b6a2eda4#diff-e1546f0e7fe0b238191ff0cb82342936

kolyshkin replied:
The whole package is deprecated (see the last commit). From what I understood from https://github.com/golang/go/wiki/Deprecated, it should be sufficient.

I guess you were reviewing the second commit. At this point, this package is still used as marking it deprecated will result in validation failure. So, I only marked it as deprecated once I switched the codebase to use moby/sys

thaJeztah commented on pkg/mount/mount.go:
```diff
@@ -1,151 +1,20 @@
-	return mount(device, target, mType, uintptr(flag), data)
-}
+// Deprecated: this package is no longer maintained and will be removed.
+// Pplease switch to github.com/moby/sys/mount (for mounting) and/or
```
Typo Pplease 

but I think we could skip the niceties and bring the message more direct;

    // Deprecated: this package is deprecated, use github.com/moby/sys/mount and github.com/moby/sys/mountinfo instead

kolyshkin replied:
I was trying to not use the word "deprecated" twice, since it appears in documentation as is.

Thanks, fixed

thaJeztah commented on pkg/mount/mountinfo.go:
```diff
@@ -1,40 +1,22 @@
+
+//nolint:golint
+type (
+	FilterFunc = mountinfo.FilterFunc
```
Same as the other PR; I think we can probably collect all these in a single "deprecated.go" file

kolyshkin replied:
done

thaJeztah commented on pkg/mount/deprecated.go:
```diff
@@ -39,15 +44,18 @@ const (
 )
 
+// Deprecated: use github.com/moby/sys/mount instead.
 //nolint:golint
```
Is the linter complaining without the `//nolint` comments? The downside of these is that they are only taken into account by our custom linters (such as golang-ci-lint), but, e.g. ignored by `golang.org/x/lint` itself, and thus would show up on https://goreportcard.com/report/github.com/containerd/containerd#golint

kolyshkin replied:
nolint:golint was added to get rid of "not documented" warnings. Maybe when "Deprecated" documentation is added, this is no longer needed. Let me try.

kolyshkin replied:
Let's see if removing `golint:nolint` works (if not, I'll revert)

kolyshkin replied:
```
[2020-03-19T18:27:31.851Z] // Deprecated: use github.com/moby/sys/mount instead.
[2020-03-19T18:27:31.851Z] ^
[2020-03-19T18:27:31.851Z] pkg/mount/deprecated.go:48:1: comment on exported type FilterFunc should be of the form "FilterFunc ..." (with optional leading article) (golint)
[2020-03-19T18:27:31.851Z] // Deprecated: use github.com/moby/sys/mountinfo instead.
[2020-03-19T18:27:31.851Z] ^
```

Nope, `golint` still complains, reverting the last change back

kolyshkin commented:
updated to latest moby/sys

cpuguy83 commented:
```
[2020-03-19T19:38:46.083Z] 

[2020-03-19T19:38:46.083Z] ---> Making bundle: cross (in bundles/cross)

[2020-03-19T19:38:46.083Z] Cross building: bundles/cross/linux/amd64

[2020-03-19T19:38:46.083Z] Building: bundles/cross/linux/amd64/dockerd-dev

[2020-03-19T19:38:46.083Z] GOOS="linux" GOARCH="amd64" GOARM=""

[2020-03-19T19:40:07.493Z] Created binary: bundles/cross/linux/amd64/dockerd-dev

[2020-03-19T19:40:07.493Z] Copying nested executables into bundles/cross/linux/amd64

[2020-03-19T19:40:07.493Z] Cross building: bundles/cross/windows/amd64

[2020-03-19T19:40:07.493Z] Building: bundles/cross/windows/amd64/dockerd-dev.exe

[2020-03-19T19:40:07.493Z] GOOS="windows" GOARCH="amd64" GOARM=""

[2020-03-19T19:40:25.565Z] # github.com/docker/docker/vendor/github.com/moby/sys/mountinfo

[2020-03-19T19:40:25.565Z] vendor/github.com/moby/sys/mountinfo/mountinfo.go:16:9: undefined: parseInfoFile
```

kolyshkin commented:
Oh shoot, this is caused by https://github.com/moby/sys/pull/6. Fixed by 

Eventually, we need to revert https://github.com/moby/moby/pull/21386, or somehow deprecate windows altogether.

kolyshkin commented:
revendored newer moby/sys to fix windows build

cpuguy83 reviewed (approved):
LGTM

kolyshkin commented:
Is there anything else that should be done here? It looks like ready to be merged from where i stand.

thaJeztah commented:
Looks like golint still expects some godoc comments to be well-formed

```bash
go get -u golang.org/x/lint/golint
golint pkg/mount/...
pkg/mount/deprecated.go:47:1: comment on exported var MergeTmpfsOptions should be of the form "MergeTmpfsOptions ..."
pkg/mount/deprecated.go:51:1: comment on exported type FilterFunc should be of the form "FilterFunc ..." (with optional leading article)
pkg/mount/deprecated.go:51:1: comment on exported type Info should be of the form "Info ..." (with optional leading article)
```

kolyshkin committed:
pkg/mount: make standalone golint happy

We do our CI via golangci-lint, which understands nolint: annotations.

A standalone linter tool, golint, does not, and it insists on
documenting these:

> pkg/mount/deprecated.go:47:1: comment on exported var MergeTmpfsOptions should be of the form "MergeTmpfsOptions ..."
> pkg/mount/deprecated.go:51:1: comment on exported type FilterFunc should be of the form "FilterFunc ..." (with optional leading article)
> pkg/mount/deprecated.go:51:1: comment on exported type Info should be of the form "Info ..." (with optional leading article)

For `MergeTmpfsOptions`, the workaround is to put it inside a
`var ( ... )` block.

For the other two warnings, we have to provide the "actual"
documentation (or something that looks like it).

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Added a separate commit to shut up a standalone-running `golint`.

thaJeztah reviewed (approved):
LGTM

perhaps the `//nolint` comments are not even needed anymore now

tao12345666333 commented on vendor/github.com/moby/sys/mountinfo/mountinfo_linux.go:
```diff
@@ -1,4 +1,6 @@
-package mount // import "github.com/docker/docker/pkg/mount"
+// +build go1.13
```
**Hello, I wonder if the build condition here has a special meaning?**

----
```
#37 0.675 Removing /build/bundles/
 #37 0.677 
 #37 0.684 ---> Making bundle: binary (in /build/bundles/binary)
 #37 0.821 Building: /build/bundles/binary-daemon/dockerd-0.0.0-20200404132736-92768b3296
 #37 0.823 GOOS="" GOARCH="" GOARM=""
 #37 33.99 # github.com/docker/docker/vendor/github.com/moby/sys/mountinfo
 #37 33.99 vendor/github.com/moby/sys/mountinfo/mountinfo.go:8:9: undefined: parseMountTable
 #37 33.99 vendor/github.com/moby/sys/mountinfo/mountinfo.go:16:9: undefined: parseInfoFile
 #37 ERROR: executor failed running [/bin/sh -c hack/make.sh binary]: runc did not terminate sucessfully
 ------
  > [build-binary 1/1] RUN --mount=type=cache,target=/root/.cache/go-build     --mount=type=bind,target=/go/src/github.com/docker/docker         hack/make.sh binary:
 ------
 failed to solve with frontend dockerfile.v0: failed to solve with frontend gateway.v0: rpc error: code = Unknown desc = failed to build LLB: executor failed running [/bin/sh -c hack/make.sh binary]: runc did not terminate sucessfully
 make[2]: *** [Makefile:188: binary] Error 1
 make[2]: Leaving directory '/builds/[MASKED]/docker-ce-packaging/docker-ce/components/engine'
 make[1]: *** [Makefile:73: static-engine] Error 2
 make[1]: Leaving directory '/builds/[MASKED]/docker-ce-packaging/static'
 make: *** [Makefile:35: static] Error 2
```

I used GitLab CI for nightly build, but the recent task failed, and the above error message was found in the log.
After checking, I found that it was caused by the build condition here. (the build job using golang 1.12)

thaJeztah replied:
Yes, this package requires Go 1.12 or up

tao12345666333 replied:
Thanks for your reply.
Go 1.12 or up? Should we change the  build condition?

thaJeztah replied:
oh, sorry, wanted to type 1.13, and mis-typed 😂

tao12345666333 replied:
I have updated my build job to use 1.13 and it works fine. 

By the way, I want to ask, what is the reason why it needs 1.13? I checked the code and did not see the features that must be supported in 1.13, I don’t know if I missed something. Thanks!
